### ISSUE TITLE
List data type field `FromJson` not getting correctly generated

### ISSUE THREAD

**Issue by Vedsaga** [jankuss/genq#41]
### Input file

```dart
import 'package:genq/genq.dart';

part 'genq_test.genq.dart';

@Genq(json: true)
class User with _$User {
  factory User({
    @JsonKey(name: 'full_name') required List<String> name,
  }) = _User;
}
```

### Current generated code

```dart

User $UserFromJson(Map<String, dynamic> json) {
  return User(
    name: List.of(json['full_name']).map((e) => e as String).toList(),
  );
}
```

As could be seen `json['full_name']` data type would be dynamic, so we can't loop over it, we would need to specifically cast it be iterable. The dart analyzer will give error

`
The argument type 'dynamic' can't be assigned to the parameter type 'Iterable<dynamic>'. dart[argument_type_not_assignable](https://dart.dev/diagnostics/argument_type_not_assignable)
`

### Expected generated code

```dart

User $UserFromJson(Map<String, dynamic> json) {
  return User(
    name: List.of(json['full_name'] as Iterable).map((e) => e as String).toList(),
  );
}
```
There were some more improvement which could be done by taking inspiration from freeze, not reporting right in this for now.

Btw, this is great initiative to improve the code gen experience for classes. Today came across the package thought to give it try, as currently when using the freeze with build running is taking ~3 min per build and can go up to ~5-8 min some times.

While macros in dart around the corner, expected to be released by early 2025 and all build runner related package expected to move to macros. Still it will good to have alternate. 

Will try my best to contribute to it and improve it.

Thank you again.

**Comment by jankuss** [comment]
Hello, sorry for the long delay. 

Likely this is caused due to the following setting in the `analysis_options.yaml`:

```yaml
 include: package:lints/recommended.yaml
 analyzer:
   language:
>    strict-casts: true
```

The generated code by genq heavily relies on passing the dynamic values along without any casting. This makes code generation simpler and faster, as we do not have to lookup the types of any method parameters during codegen. While we could cast the `json['full_name']` as `Iterable` in your example, the above rule would apply in other cases as well where we would not really know the type to cast to.

For example, a user is able to specify a custom `fromJson` method, where he can basically specify an arbitrary type as the parameter. `genq` does not have the ability to figure out this parameter type, so we cannot generate the correct casting. Thus, we would need to broaden the type to something more general, like `Object`, and move the actual casting into the functions (in this example, the user would need to do the casting in his custom `fromJson`).

---

**TLDR:** I would suggest you exclude `**.genq.dart` from analysis. `genq` won't have to change its architecture, and you will be able to happily generate the code.

```
include: package:lints/recommended.yaml
analyzer:
  exclude:
    - '**.genq.dart'
```

I think we should document this in the README though, as this is a common enough setting. You shouldn't be the only one bumping into this problem.

**Comment by Vedsaga** [comment]
Thank you for the response, and this is great to know that there is work around. 

While earlier was trying to migrate from the old project to new. Due to earlier shared case, didn't proceed with that plan.

In couple of months planing to start new project in that project will certainly going to give it another try.
